src/future.ml

```diff
@@ -319,6 +334,75 @@ module Scheduler = struct
       | Terminal -> s
       | File fn | Opened_file { filename = fn; _ } -> sprintf "%s 2> %s" s fn
 
+  let pp_purpose ppf = function
+  | Internal_job ->
+     Format.fprintf ppf "(internal)"
+  | Build_job targets ->
+     let rec split_paths targets_acc ctxs_acc = function
+       | [] -> List.rev targets_acc, List.sort_uniq ~cmp:compare ctxs_acc
+       | path :: rest ->
+          match Path.extract_build_context path with
+          | None ->
+             split_paths (Path.to_string path :: targets_acc) ctxs_acc rest
+          | Some ("default", filename) ->
+             split_paths (Path.to_string filename :: targets_acc) ctxs_acc rest
+          | Some (".aliases", filename) ->
+             let ctxs_acc, filename = match Path.extract_build_context filename with
+               | None -> ctxs_acc, Path.to_string filename
+               | Some (ctx, fn) ->
+                  let strip_digest fn =
+                    let fn = Path.to_string fn in
+                    match String.rsplit2 fn ~on:'-' with
+                    | None -> fn
+                    | Some (name, digest) ->
+                       match Digest.from_hex digest with
+                       | _ -> name
+                       | exception (Invalid_argument _) -> fn in
+                  let ctxs_acc =
+                    if ctx = "default" then ctxs_acc else ctx :: ctxs_acc in
+                  ctxs_acc, strip_digest fn in
+             split_paths (("alias " ^ filename) :: targets_acc) ctxs_acc rest
+          | Some (ctx, filename) ->
+             split_paths (Path.to_string filename :: targets_acc) (ctx :: ctxs_acc) rest in
+     let target_names, contexts = split_paths [] [] targets in
+     let rec group_by_ext = function
```

Comment:
This would probably be clearer with a combination of `Filename.split_ext` and `String_map.of_alist_multi`, but it's not very important, it can stay like this.